src/deploy/functions/params.ts

```diff
@@ -267,83 +396,271 @@ export async function resolveParams(
     case "text":
     default:
-      let prompt = `Enter a value for ${param.label || param.name}:`;
+      prompt = `Enter an integer value for ${param.label || param.name}:`;
       if (param.description) {
         prompt += ` \n(${param.description})`;
       }
-      let res: number;
-      while (true) {
-        res = await promptOnce({
-          name: param.name,
-          type: "number",
-          default: resolvedDefault,
-          message: prompt,
-        });
-        if (Number.isInteger(res)) {
-          return res;
+      return promptText<number>(prompt, param.input, resolvedDefault, (res: string) => {
+        if (isNaN(+res)) {
+          return { message: `"${res}" could not be converted to a number.` };
+        }
+        if (res.includes(".")) {
+          return { message: `${res} is not an integer value.` };
         }
-        logger.error(`${param.label || param.name} must be an integer; retrying...`);
+        return +res;
+      });
+  }
+}
+
+async function promptResourceString(
+  prompt: string,
+  input: ResourceInput,
+  projectId: string,
+  resolvedDefault?: string
+): Promise<string> {
+  const notFound = new FirebaseError(`No instances of ${input.resource.type} found.`);
+  switch (input.resource.type) {
+    case "storage.googleapis.com/Bucket":
+      const buckets = await listBuckets(projectId);
+      if (buckets.length === 0) {
+        throw notFound;
       }
+      const forgedInput: SelectInput<string> = {
+        type: "select",
+        select: buckets.map((bucketName: string): SelectOptions<string> => {
+          return { label: bucketName, value: bucketName };
+        }),
+      };
+      return promptSelect<string>(prompt, forgedInput, resolvedDefault, (res: string) => res);
+    default:
+      logger.warn(
+        `Warning: unknown resource type ${input.resource.type}; defaulting to raw text input...`
+      );
+      return promptText<string>(prompt, { type: "text" }, resolvedDefault, (res: string) => res);
+  }
+}
+
+type retryInput = { message: string };
+async function promptText<T extends RawParamValue>(
+  prompt: string,
+  input: TextInput<T>,
+  resolvedDefault: T | undefined,
+  converter: (res: string) => T | retryInput
+): Promise<T> {
+  const res = await promptOnce({
+    type: "input",
+    default: resolvedDefault,
+    message: prompt,
+  });
+  if (input.validationRegex) {
+    const userRe = new RegExp(input.validationRegex);
+    if (!userRe.test(res)) {
+      logger.error(
+        input.validationErrorMessage ||
+          `Input did not match provided validator ${userRe.toString()}, retrying...`
+      );
+      return promptText<T>(prompt, input, resolvedDefault, converter);
+    }
   }
+  const converted = converter(res);
+  if (typeof converted === "object") {
+    logger.error(converted.message);
+    return promptText<T>(prompt, input, resolvedDefault, converter);
+  }
```

Comment:
hmm this feel overlapping with validationRegex and a bit confusing to understand without looking at how this argument is actually used by the caller.

One alternative would be to define `promptNumber` or `promptBoolean` to specifically cover those cases.

(no action required if you disagree)